_posts/2015-11-11-full-stack-es6-with-webpack-and-babel.md

```diff
@@ -0,0 +1,81 @@
+
+Under the hood this means aggregating and manipulating JSON data from various REST resources, and providing a live preview of the result in the layout edit mode. Both are tasks JavaScript is really good at, so my team does **JS all the way** from the (Mongo) database through the (Node) server down to the browser.
+
+## JavaScript modules revisited
+One of our pains as JavaScript developers has always been **the lack of a module system**. Node.js introduced the _CommonJS_ approach consisting of `module.exports` and `require('module')`, which works well on the server but not on the client. That's because in the browser, we have network latency, and thus loading a module is an asynchronous operation. When we started our project, _RequireJS_ solved this for us as elegant as it can get without introducing a compile step, using its _AMD (asynchronous module definition)_ syntax consisting of `define(['module'], function (module) {...});`.
+
+But we're writing _isomorphic_ JavaScript, which is fancy speaking for running (parts of) the exact same code on client and server alike. If you're doing this in combination with RequireJS, you can easily get a headache from looking at statements like `var require = require('requirejs');`.
+
+**Enter _Webpack_.** Just like its cousin _Browserify_, it gets around the asynchronous loading issue by introducing a build step that compiles related modules into _bundles_. As a bonus, it offers a variety of pluggable _loaders_ that can transform things before putting them into the bundle. In our case this comes in handy for including _React JSX_, _Less_ CSS and _Dust_ template files.
+
+## Suddenly, ES6
+But the biggest benefit comes with the **_Babel_** Webpack loader. Using the widespread _Babel_ transpiler, formerly known as _6to5_ because it rewrites ES6 (and ES7) code to the ES5 language level every browser understands, we can now use most standardised ES6 features right away! Coming back to our module challenge, this adds to the list of module systems supported by Webpack (which comes with CommonJS and AMD support built in) **the canonical module system of ECMAScript 6**.
+
+Let's have a look at some of our ES6 module code:
+{% highlight javascript %}
+import {DragSource} from '../../utils/DragAndDrop';
+import Immutable from 'immutable';
+import ImmutablePropTypes from 'react-immutable-proptypes';
+import React from 'react';
+
+export default class ImagePlugin extends React.Component {
+  constructor() {
+    super();
+{% endhighlight %}
+
+Much leaner than good ol' RequireJS syntax, isn't it?
+The alert reader might notice some more ES6 novelties here: Yup, ES6 introduces the `class`, `extends`, and `super` keywords as syntactic sugar over JavaScript's prototypal inheritance system, something you might already have been using in other ECMAScript dialects like TypeScript (or, in my case, ActionScript 2).
```

Comment:
something you might have already been using in other ECMAScript dialects like TypeScript (or, in my case, ActionScript 2).